WebCryptoAPI/sign_verify/ecdsa.js

```diff
@@ -0,0 +1,391 @@
+                    .then(function(is_verified) {
+                        assert_true(is_verified, "Round trip verification works");
+                        return signature;
+                    }, function(err) {
+                        assert_unreached("verify error for test " + vector.name + ": " + err.message + "'");
+                    });
+                }, function(err) {
+                    assert_unreached("sign error for test " + vector.name + ": '" + err.message + "'");
+                });
+            }, vector.name + " round trip");
+
+        }, function(err) {
+            // We need a failed test if the importVectorKey operation fails, so
+            // we know we never tested signing or verifying
+            promise_test(function(test) {
+                assert_unreached("importVectorKeys failed for " + vector.name + ". Message: ''" + err.message + "''");
+            }, "importVectorKeys step: " + vector.name + " round trip");
+        });
+
+        all_promises.push(promise);
+    });
+
+    // Test signing with the wrong algorithm
+    testVectors.forEach(function(vector) {
+        // Want to get the key for the wrong algorithm
+        var promise = subtle.generateKey({name: "HMAC", hash: "SHA-1"}, false, ["sign", "verify"])
+        .then(function(wrongKey) {
+            var algorithm = {name: vector.algorithmName, hash: vector.hashName};
+            return importVectorKeys(vector, ["verify"], ["sign"])
+            .then(function(vectors) {
+                promise_test(function(test) {
+                    var operation = subtle.sign(algorithm, wrongKey, vector.plaintext)
+                    .then(function(signature) {
+                        assert_unreached("Signing should not have succeeded for " + vector.name);
+                    }, function(err) {
+                        assert_equals(err.name, "InvalidAccessError", "Should have thrown InvalidAccessError instead of '" + err.message + "'");
+                    });
+
+                    return operation;
+                }, vector.name + " signing with wrong algorithm name");
+
+            }, function(err) {
+                // We need a failed test if the importVectorKey operation fails, so
+                // we know we never tested verification.
+                promise_test(function(test) {
+                    assert_unreached("importVectorKeys failed for " + vector.name + ". Message: ''" + err.message + "''");
+                }, "importVectorKeys step: " + vector.name + " signing with wrong algorithm name");
+            });
+        }, function(err) {
+            promise_test(function(test) {
+                assert_unreached("Generate wrong key for test " + vector.name + " failed: '" + err.message + "'");
+            }, "generate wrong key step: " + vector.name + " signing with wrong algorithm name");
+        });
+
+        all_promises.push(promise);
+    });
+
```

Comment:
- R and S are the wrong length - i.e. 20 bytes for P-521
Tested with wrong length signature

- Use correct curve, but the wrong hash algorithm - truncation questions
Tested now.

- Correct curve but hash algorithm is not a hash algorithm
Tested now


On Tue, Jun 21, 2016 at 1:22 AM, Jim Schaad <notifications@github.com>
wrote:

> In WebCryptoAPI/sign_verify/ecdsa.js
> <https://github.com/w3c/web-platform-tests/pull/3183#discussion_r67809593>
> :
>
> > +            }, function(err) {
> > +                // We need a failed test if the importVectorKey operation fails, so
> > +                // we know we never tested verification.
> > +                promise_test(function(test) {
> > +                    assert_unreached("importVectorKeys failed for " + vector.name + ". Message: ''" + err.message + "''");
> > +                }, "importVectorKeys step: " + vector.name + " signing with wrong algorithm name");
> > +            });
> > +        }, function(err) {
> > +            promise_test(function(test) {
> > +                assert_unreached("Generate wrong key for test " + vector.name + " failed: '" + err.message + "'");
> > +            }, "generate wrong key step: " + vector.name + " signing with wrong algorithm name");
> > +        });
> > +
> > +        all_promises.push(promise);
> > +    });
> > +
>
> Additional Error cases:
> R and S are the wrong length - i.e. 20 bytes for P-521
> Use correct curve, but the wrong hash algorithm - truncation questions
> Correct curve but hash algorithm is not a hash algorithm
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3183/files/b16dcea8a391dfaa70a98ed41139c24f2c9a168d#r67809593>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAE_N4qjC0s9GDJSIsTgVFJ_3iTtqvgEks5qN3UhgaJpZM4I2ltT>
> .
>